### ISSUE TITLE
Auto closing on '{' #421

### ISSUE THREAD

**Issue by AstrickHarren** [yzhang-gh/vscode-markdown#846]
## Proposal

It seems the markdown language somehow disables auto closing for `{` when the cursor is placed before any text (expect whitespace). The most prominent case is before a `$`. As an example, if I try to type `{` in the cursor positions below (denoted by `|`), 

```latex
$here the auto closing works well \mathbf| $, $but not here \mathbf|$ 
```

this is what happens

```latex
$here the auto closing works well \mathbf{} $, $but not here \mathbf{$
```

However, if I change the setting `editor: auto closing bracket` from `language defined` to `always`, both of the above cases work well. That's way I'm suggesting the language extension to improve it.  

One may argue to keep a white space before the closing `$`, as in the first case of the example above, but this disables the syntax highlighting for `text.katex` as it is not matched by the grammar.

**Comment by Lemmingh** [comment]
Currently, this extension does **NOT** provide language configuration, including autoclosing. **It is contributed by VS Code itself**:
<https://github.com/microsoft/vscode/blob/700ee37a240d3789dbc093f9e32758f2295ea694/extensions/markdown-basics/language-configuration.json>

Although, we can also define a language configuration to overwrite the built-it one to meet this request, I am not sure **whether we should do it**.

@yzhang-gh Your opinion? Seems [you were interested in it](https://github.com/yzhang-gh/vscode-markdown/issues/421#issuecomment-485817035), but [VS Code does not](https://github.com/microsoft/vscode/issues/74046#issuecomment-494491737).

---

> the markdown language somehow disables auto closing for ...

Yes, it is the [default behavior](https://code.visualstudio.com/api/language-extensions/language-configuration-guide#autoclosing-before):

> By default, VS Code only autocloses pairs if there is whitespace right after the cursor.

**Comment by Lemmingh** [comment]
Wait, let's clarify.

This request is to:

Enable autoclosing for `{` in more scenarios in

* Markdown.
* math area only.

@AstrickHarren
which do you mean? Thank you.

**Comment by yzhang-gh** [comment]
> this is what happens
> 
> ```
> $here the auto closing works well \mathbf{} $, $but not here \mathbf{$
> ```

As @Lemmingh said, these auto-closing pairs are defined by VS Code itself (including `{}` _already_).

The problem you described is VS Code doesn't automatically close a pair if it is right before a `$`. You can try `([<` etc.
You can also see this behavior elsewhere, e.g., `foo|()` in `python`.

I don't know when the VS Code will close the pair (or not). But what is for sure is it is out of our (the extension's) control...

> @yzhang-gh Your opinion? Seems [you were interested in it](https://github.com/yzhang-gh/vscode-markdown/issues/421#issuecomment-485817035), but [VS Code does not](https://github.com/microsoft/vscode/issues/74046#issuecomment-494491737).

`{}` is already in the "auto-closing pairs" (although not in "surrounding pairs"). For me, It is good enough.

**Comment by AstrickHarren** [comment]
So I mentioned that I actually kind of fixed the problem by setting the `editor: auto closing bracket` from `language defined` to `always`. This is curious though, if VS Code defines two sets of auto closing, one for `markdown`, and one for `always`.

**Comment by yzhang-gh** [comment]
> > the markdown language somehow disables auto closing for ...
> 
> Yes, it is the [default behavior](https://code.visualstudio.com/api/language-extensions/language-configuration-guide#autoclosing-before):
> 
> > By default, VS Code only autocloses pairs if there is whitespace right after the cursor.

Just found this (`autoCloseBefore`) 😮, which means we can help with this scenario (I mean right before `$`). But I'm hesitant whether we should override the default language configuration.

**Comment by Lemmingh** [comment]
Good. You eventually noticed it. 😄

If only for math areas, it can be quite easy: We can [take advantage of other extensions](https://github.com/users/Lemmingh/projects/2#card-45404102).

---

> when the VS Code will close the pair

1. By default, only if before whitespace.
2. Then, extensions can contribute language configurations to overwrite it.
3. Finally, users can change `editor.autoClosing*` settings to overwrite it again.

**Comment by yzhang-gh** [comment]
> If only for math areas, it can be quite easy: We can [take advantage of other extensions](https://github.com/users/Lemmingh/projects/2#card-45404102).

I like this way. But (if my memory serves me right) it requires a kind of language client?

**Comment by Lemmingh** [comment]
> it requires a kind of language client?

No, [`embeddedLanguages`](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#embedded-languages) is just a **very basic scope mapping**, and enables **limited [declarative language features](https://github.com/microsoft/vscode/blob/d310cba7315b35b5d6a829360e56d0ce2720455a/src/vs/editor/common/modes/languageConfiguration.ts#L26)**. It's still [fully handled by VS Code](https://github.com/microsoft/vscode/issues/98621#issuecomment-636926460), and behaves a bit [**mysteriously** sometimes](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide#injection-grammars).

We only need to add about two lines in `package.json` to adopt it. But I have to admit that this approach **cannot solve the problem reliably**.

---

If we're going to make things **robust** and rich, a language service for LaTeX, a proper server, and a corresponding client are indeed required. VS Code's HTML Language Features gives a good example of auto closing.

**Comment by yzhang-gh** [comment]
Then I think it doesn't add much values to adopt the embedded-language solution. Our current syntax highlight and math function completions work reasonably well (especially as KaTeX is different from LaTeX).

**Comment by Lemmingh** [comment]
> it doesn't add much values to adopt the embedded-language solution

I agree.

To clarify, although the `embeddedLanguages` contribution point is introduced in "**Syntax Highlight Guide**", it's actually to **ask VS Code to load language configuration and snippets**. And from my experience, only **comment toggling** and **snippets** are always properly loaded. 😂

---

> KaTeX is different from LaTeX

Right. You remind me. KaTeX is a subset of LaTeX. Users would be confused, if snippets for LaTeX appeared in our math areas. **This a critical reason that we should not mark them as LaTeX.**

**Comment by shreyasminocha** [comment]
Related: not auto-closing `<` and recognizing `{}` as surrounding pairs in math contexts.

**Comment by yzhang-gh** [comment]
Thanks for the information. But why can't I find this line there? It doesn't seem to exist in the VS Code source code 🤔

https://github.com/microsoft/vscode/blob/main/extensions/markdown-basics/language-configuration.json

**Comment by yfzhao20** [comment]
哦，当然没有😂因为这一段是我从latex workshop抄过来的，为的就是让括号能够自动闭合。

直接把这一句代码加上就行，目前还没发现什么bug

**Comment by Lemmingh** [comment]
Found something interesting:
<https://github.com/microsoft/vscode/issues/133397>

I guess [my observation](https://github.com/yzhang-gh/vscode-markdown/issues/846#issuecomment-733586401) was correct.

**Comment by Lemmingh** [comment]
> [not auto-closing `<`](https://github.com/yzhang-gh/vscode-markdown/issues/846#issuecomment-783906988)

Let's see if VS Code can come up with a solution:
<https://github.com/microsoft/vscode/issues/131331>

---

I don't think the problems can be **correctly** solved until VS Code tells us how to **programmatically (dynamically) define embedded language scope**.

Although, the [Semantic Highlighting Guide](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-classification) says:

> Optionally, the classification can also name a language, if the token is part of an embedded language.

no further explanation exists.

There are also a bunch of [confusing discussions](https://github.com/microsoft/vscode/issues?q=label%3Asemantic-tokens+embedded+language):

* <https://github.com/microsoft/vscode/issues/81558>

* <https://github.com/microsoft/vscode/issues/86329>

* <https://github.com/microsoft/vscode/issues/113640#issuecomment-754061490>
